File: contesst10_khoangcach.c
#include <stdio.h>
#include <stdlib.h>

#define INF 1000000000

int n, k, p;
int people[1000], cakes[2000];

int cmp(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

int can_assign(int max_time) {
    int i = 0, j = 0;
    while (i < n && j < k) {
        if (abs(people[i] - cakes[j]) + abs(cakes[j] - p) <= max_time) {
            i++; 
        }
        j++; 
    }
    return i == n; 
}

int main() {

    scanf("%d %d %d", &n, &k, &p);
    for (int i = 0; i < n; i++) scanf("%d", &people[i]);
    for (int i = 0; i < k; i++) scanf("%d", &cakes[i]);

    qsort(people, n, sizeof(int), cmp);
    qsort(cakes, k, sizeof(int), cmp);

    int left = 0, right = INF, ans = INF;
    while (left <= right) {
        int mid = (left + right) / 2;
        if (can_assign(mid)) {
            ans = mid;
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    printf("%d\n", ans);
    return 0;
}

File: contesst10_khoangcach.py
def can_assign(T, persons, cakes, p):
    n = len(persons)
    k = len(cakes)
    j = 0  
    for a in persons:

        if T < abs(a - p):
            return False

        L = (a + p - T) / 2
        R = (a + p + T) / 2

        while j < k and cakes[j] < L:
            j += 1

        if j < k and cakes[j] <= R:
            j += 1
        else:
            return False
    return True

def min_time(persons, cakes, p):
    persons.sort()
    cakes.sort()
    low = 0
    high = max(max(abs(p - a) for a in persons), max(abs(p - b) for b in cakes)) * 2
    ans = high
    while low <= high:
        mid = (low + high) // 2
        if can_assign(mid, persons, cakes, p):
            ans = mid
            high = mid - 1
        else:
            low = mid + 1
    return ans

if __name__ == '__main__':

    persons = [10, 25, 40]
    cakes = [8, 15, 20, 35, 50]
    p = 30
    print(min_time(persons, cakes, p))